File: 1644-maximum-number-of-non-overlapping-substrings/maximum-number-of-non-overlapping-substrings.java
import java.util.*;

class Solution {
    public List<String> maxNumOfSubstrings(String s) {
        int n = s.length();
        int[] first = new int[26];
        int[] last = new int[26];
        Arrays.fill(first, -1);
        Arrays.fill(last, -1);

        for (int i = 0; i < n; i++) {
            int ch = s.charAt(i) - 'a';
            if (first[ch] == -1) {
                first[ch] = i;
            }
            last[ch] = i;
        }

        List<int[]> intervals = new ArrayList<>();

        for (int i = 0; i < 26; i++) {
            if (first[i] == -1) continue;

            int start = first[i];
            int end = last[i];
            boolean valid = true;

            for (int j = start; j <= end; j++) {
                int ch = s.charAt(j) - 'a';
                if (first[ch] < start) {
                    valid = false;
                    break;
                }
                end = Math.max(end, last[ch]);
            }

            if (valid) {
                intervals.add(new int[]{start, end});
            }
        }

        // Sort intervals by end index
        intervals.sort((a, b) -> Integer.compare(a[1], b[1]));

        List<String> res = new ArrayList<>();
        int prevEnd = -1;

        for (int[] interval : intervals) {
            if (interval[0] > prevEnd) {
                res.add(s.substring(interval[0], interval[1] + 1));
                prevEnd = interval[1];
            }
        }

        return res;
    }
}

File: 1644-maximum-number-of-non-overlapping-substrings/maximum-number-of-non-overlapping-substrings.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<string> maxNumOfSubstrings(string s) {
        int n = s.length();
        vector<int> first(26, -1), last(26, -1);

        for (int i = 0; i < n; i++) {
            int ch = s[i] - 'a';
            if (first[ch] == -1) {
                first[ch] = i;
            }
            last[ch] = i;
        }

        vector<pair<int, int>> intervals;

        for (int i = 0; i < 26; i++) {
            if (first[i] == -1) continue;

            int start = first[i];
            int end = last[i];
            bool valid = true;

            for (int j = start; j <= end; j++) {
                int ch = s[j] - 'a';
                if (first[ch] < start) {
                    valid = false;
                    break;
                }
                end = max(end, last[ch]);
            }

            if (valid) {
                intervals.push_back({start, end});
            }
        }

        // Sort intervals by end index
        sort(intervals.begin(), intervals.end(), [](const pair<int, int>& a, const pair<int, int>& b) {
            return a.second < b.second;
        });

        vector<string> res;
        int prevEnd = -1;

        for (const auto& interval : intervals) {
            if (interval.first > prevEnd) {
                res.push_back(s.substr(interval.first, interval.second - interval.first + 1));
                prevEnd = interval.second;
            }
        }

        return res;
    }
};